core/crypto/src/signature.rs

```diff
@@ -556,6 +558,83 @@ impl<'de> serde::Deserialize<'de> for SecretKey {
 #[derive(Clone)]
 pub struct Secp256K1Signature([u8; 65]);
 
+impl Secp256K1Signature {
+    fn check_v(&self) -> Result<u8, crate::errors::ParseSignatureError> {
+        match self.0[64] {
+            27 | 28 => Ok(self.0[64] - 27),
+            _ => Err(crate::errors::ParseSignatureError::InvalidData {
+                error_message: "invalid `v` value".to_string(),
+            }),
+        }
+    }
+
+    pub fn check_signature_values(&self) -> bool {
+        let mut r_bytes = [0u8; 32];
+        r_bytes.copy_from_slice(&self.0[0..32]);
+        let r = U256::from(r_bytes);
+
+        let mut s_bytes = [0u8; 32];
+        s_bytes.copy_from_slice(&self.0[32..64]);
+        let s = U256::from(s_bytes);
+
+        let v = match self.check_v() {
+            Ok(v) => v,
+            Err(_e) => return false,
+        };
+
+        let big_one = U256::one();
+        if r < big_one || s < big_one {
+            return false;
+        }
+
+        let secp256k1_n = U256::from([
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xfe, 0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b, 0xbf, 0xd2, 0x5e, 0x8c,
+            0xd0, 0x36, 0x41, 0x41,
+        ]);
+
+        let secp256k1_half_n = secp256k1_n / U256::from(2);
+        // Reject upper range of s values (ECDSA malleability)
+        if s > secp256k1_half_n {
+            return false;
+        }
+
+        r < secp256k1_n && s < secp256k1_n && (v == 0 || v == 1)
```

Comment:
```suggestion
        r < secp256k1_n && s < secp256k1_n
```

The last clause is dead code which is always true -- this check is already done by `check_v`. I guess, `assert!(v == 0 || v == 1)` could be added for readability?

Reply:
Good eye, I just took that from original reference material. You are right, should be fine.

Fixed https://github.com/near/nearcore/pull/4380/commits/0acaa6700500576bd116b6fb8bfa6c0b47833838